File: src/main/java/linq/collections/enumerable/IterableEnumerable.java
package linq.collections.enumerable;

import linq.Enumerable;
import linq.Enumerator;
import linq.collections.enumerator.IterableEnumerator;

import java.util.Iterator;

public class IterableEnumerable<T> extends Enumerable<T> {
    private final Iterable<T> iterable;

    public IterableEnumerable(Iterable<T> iterable) {
        this.iterable = iterable;
    }

    @Override
    public Enumerator<T> enumerator() {
        return new IterableEnumerator<>(iterable.iterator());
    }

    @Override
    public Iterator<T> iterator() {
        return iterable.iterator();
    }
}

File: src/main/kotlin/linq/collections/enumerable/IterableEnumerable.kt
package linq.collections.enumerable

import linq.Enumerable
import linq.Enumerator
import linq.collections.enumerator.IterableEnumerator

class IterableEnumerable<T>(private val iterable: Iterable<T>) : Enumerable<T>() {
    override fun enumerator(): Enumerator<T> {
        return IterableEnumerator(iterable)
    }

    override fun iterator(): Iterator<T> {
        return iterable.iterator()
    }
}